src/mapstorage/XmlMapStorage.cpp

```diff
@@ -41,41 +198,426 @@ void XmlMapStorage::newData()
+
+    const QXmlStreamAttributes attrs = stream.attributes();
+    const QStringView idstr = attrs.value("id");
+    const RoomId roomId = loadRoomId(stream, idstr);
+    if (m_loadedRooms.count(roomId) != 0) {
+        throwErrorFmt(stream, "duplicated room id \"%1\"", idstr.toString());
+    }
+    room.setId(roomId);
+    if (attrs.value("uptodate") == "false") {
+        room.setOutDated();
+    } else {
+        room.setUpToDate();
+    }
+    room.setName(RoomName{attrs.value("name").toString()});
+
+    ExitsList exitList;
+    RoomLoadFlags loadFlags;
+    RoomMobFlags mobFlags;
+
+    while (stream.readNextStartElement() && !stream.hasError()) {
+        const std::u16string_view name = as_u16string_view(stream.name());
+        if (name == "align") {
+            room.setAlignType(loadEnum<RoomAlignEnum>(stream));
+        } else if (name == "contents") {
+            room.setContents(RoomContents{loadString(stream)});
+        } else if (name == "coord") {
+            room.setPosition(loadCoordinate(stream));
+        } else if (name == "description") {
+            room.setDescription(RoomDesc{loadString(stream)});
+        } else if (name == "exit") {
+            loadExit(stream, exitList);
+        } else if (name == "light") {
+            room.setLightType(loadEnum<RoomLightEnum>(stream));
+        } else if (name == "loadflag") {
+            loadFlags |= loadEnum<RoomLoadFlagEnum>(stream);
+        } else if (name == "mobflag") {
+            mobFlags |= loadEnum<RoomMobFlagEnum>(stream);
+        } else if (name == "note") {
+            room.setNote(RoomNote{loadString(stream)});
+        } else if (name == "portable") {
+            room.setPortableType(loadEnum<RoomPortableEnum>(stream));
+        } else if (name == "ridable") {
+            room.setRidableType(loadEnum<RoomRidableEnum>(stream));
+        } else if (name == "sundeath") {
+            room.setSundeathType(loadEnum<RoomSundeathEnum>(stream));
+        } else if (name == "terrain") {
+            room.setTerrainType(loadEnum<RoomTerrainEnum>(stream));
+        } else {
+            qWarning().noquote().nospace()
+                << "At line " << stream.lineNumber() << ": ignoring unexpected XML element <"
+                << stream.name() << "> inside <room id=\"" << idstr << "\">";
+        }
+        skipXmlElement(stream);
+    }
+    room.setExitsList(exitList);
+    room.setLoadFlags(loadFlags);
+    room.setMobFlags(mobFlags);
+
+    m_loadedRooms.emplace(std::move(roomId), std::move(sharedroom));
+}
+
+// convert string to RoomId
+RoomId XmlMapStorage::loadRoomId(QXmlStreamReader &stream, const QStringView idstr)
+{
+    bool fail = false;
+    const RoomId id{conv.toInteger<uint32_t>(idstr, fail)};
+    // convert number back to string, and compare the two:
+    // if they differ, room ID is invalid.
+    if (fail || idstr != roomIdToString(id)) {
+        throwErrorFmt(stream, "invalid room id \"%1\"", idstr.toString());
```

Comment:
Ok, I'll make marker position mandatory.

About the `fail` flag:
I did not want to throw an exception from `XmlMapStorage::Converter::toInteger<T>` because it has no context about the meaning of the number it is parsing. If it threw an exception, a caller would need to catch it and throw another exception with a more descriptive message - and forgetting to do that is easy.

Passing a `bool &fail` to `XmlMapStorage::Converter::toInteger<T>` acts as a reminder for the caller "I should check what's there after `toInteger` returns.

If you really don't like as it looks now, I can change `toInteger` to throw on non-numeric input, and wrap it everywhere with `catch (/**/) { throw /**/ }`

Reply:
marker pos1 is now mandatory

Reply:
I also updated `saveMarker` to omit pos2 if it's equal to pos1,
and updated `loadMarker` to set pos2 = pos1 if the former is not present in the XML file.